Add HTTP_HOST to fastcgi params

teward opened the pull request:
Downstream origin of request: https://salsa.debian.org/nginx-team/nginx/-/issues/21

### Proposed changes

It was indicated in a downstream 'issue' in the Debian Salsa repository for NGINX that:

> When using HTTP/3, the Host header is not sent anymore by the client, it always and only uses the :authority pseudo-header. Thus, fastcgi never receives it, and some PHP applications rely on it because it's mandatory on HTTP/1.1.
> 
> I propose to use fastcgi_param HTTP_HOST $host because as mentioned here (https://trac.nginx.org/nginx/ticket/2468#comment:1) it will use the real computed Host value from nginx.

Because changes to nginx open source are now able to be done via git PRs, etc. I have 'upstreamed' this issue to here as a PR.

MarcoMarcoaldi commented:
Please check the fix in this new pull request and patch :

https://github.com/nginx/nginx/pull/917

ac000 committed:
Add HTTP_HOST parameter.

This ensures that a HTTP_HOST environment variable is always set
regardless of HTTP protocol version used.

Link: <https://salsa.debian.org/nginx-team/nginx/-/issues/21>
Co-authored-by: Thomas Ward <teward@thomas-ward.net>

ac000 commented:
@teward

So it seems this can now be done properly now that
[Added $request_port and $is_request_port variables.](https://github.com/nginx/nginx/commit/c8c7beb96f61e2251abbc345357116131cf91c22)
has been merged.

So the actual entry would be

```
fastcgi_param  HTTP_HOST          $host$is_request_port$request_port;
```

I guess this should also go into `conf/fastcgi.conf` and also similar
entries in `conf/scgi_params` & `conf/uwsgi_params`.

ac000 commented:
Rebased with master

```
$ git range-diff a42220319...e19e7a809
  -:  --------- >   1:  144778aee Version bump.
  -:  --------- >   2:  3f6d94d88 QUIC: prevent deleted stream frame retransmissions.
  -:  --------- >   3:  b394d44cf Configure: MSVC compatibility with PCRE2 10.43.
  -:  --------- >   4:  f45c2707e Updated security policy to include disclosure details.
  -:  --------- >   5:  ebd18ec18 SSL: disabled TLSv1 and TLSv1.1 by default.
  -:  --------- >   6:  db6870e06 Upstream: re-resolvable servers.
  -:  --------- >   7:  9fe119b43 Upstream: construct upstream peers from DNS SRV records.
  -:  --------- >   8:  1524c5e3f Core: inheritance of non-reusable shared memory zones.
  -:  --------- >   9:  5ebe7a412 Upstream: pre-resolve servers on reload.
  -:  --------- >  10:  ea4654550 Upstream: per-upstream resolver.
  -:  --------- >  11:  29aec5720 Upstream: copy upstream zone DNS valid time during config reload.
  -:  --------- >  12:  ea15896c1 SSL: fixed MSVC compilation after ebd18ec1812b.
  -:  --------- >  13:  a5e152b3d FastCGI: fixed create_loc_conf comments after 05b1a8f1e.
  -:  --------- >  14:  1ac6a1858 SCGI: added create_loc_conf comments.
  -:  --------- >  15:  d10bf73eb Uwsgi: added create_loc_conf comments.
  -:  --------- >  16:  36ca44f26 Fixed link to contributing guidelines.
  -:  --------- >  17:  7cd60cd47 On DragonFly BSD 5.8+, TCP_KEEPIDLE and TCP_KEEPINTVL are in secs.
  -:  --------- >  18:  0ebc3242d SSL: error message default in object caching API.
  -:  --------- >  19:  cb1857407 Fixed missing double quote.
  -:  --------- >  20:  6ec099a37 Mp4: fixed handling an empty run of chunks in stsc atom.
  -:  --------- >  21:  d1a02451c Mp4: unordered stsc chunks error for the final chunk.
  -:  --------- >  22:  569948aa1 Mp4: prevent chunk index underflow.
  -:  --------- >  23:  476d6526b SSL: a new macro to set default protocol versions.
  -:  --------- >  24:  0864cca4d QUIC: prevented BIO leak in case of error.
  -:  --------- >  25:  a448dd52e Upstream: disallow empty path in proxy_store and friends.
  -:  --------- >  26:  9a025219f QUIC: constified nonce parameter of crypto functions.
  -:  --------- >  27:  3f755b5a9 QUIC: got rid of memory copy when initializing constant values.
  -:  --------- >  28:  b2a67d261 Realip: allowed square brackets with portless IPv6 address.
  -:  --------- >  29:  ce88b1712 Mail: handling of LOGIN IMAP command untagged response.
  -:  --------- >  30:  e7bd25574 nginx-1.27.3-RELEASE
  -:  --------- >  31:  e28ef42b9 Version bump.
  -:  --------- >  32:  930caed3b QUIC: fixed client request timeout in 0-RTT scenarios.
  -:  --------- >  33:  c73fb273a Updated security policy to clarify experimental features.
  -:  --------- >  34:  a52ba8ba0 QUIC: ignore version negotiation packets.
  -:  --------- >  35:  e3a9b6ad0 QUIC: fixed accessing a released stream.
  -:  --------- >  36:  febe6e728 Year 2025.
  -:  --------- >  37:  57d54fd92 Gzip: compatibility with recent zlib-ng 2.2.x versions.
  -:  --------- >  38:  47f862ffa Slice filter: log the expected range in case of range error.
  -:  --------- >  39:  8311e14ae SSL: object cache inheritance from the old configuration cycle.
  -:  --------- >  40:  7677d5646 SSL: encrypted certificate keys are exempt from object cache.
  -:  --------- >  41:  0e756d67a SSL: caching certificates and certificate keys with variables.
  -:  --------- >  42:  4b96ad14f SSL: cache revalidation of file based dynamic certificates.
  -:  --------- >  43:  454ad0ef3 Upstream: caching certificates and certificate keys with variables.
  -:  --------- >  44:  5d5d9adcc SSL: avoid using mismatched certificate/key cached pairs.
  -:  --------- >  45:  5ab4f32e9 Upstream: fixed --with-compat build without SSL, broken by 454ad0e.
  -:  --------- >  46:  64d0795ac QUIC: added missing casts in iov_base assignments.
  -:  --------- >  47:  e71520222 Configure: fixed --with-libatomic=DIR with recent libatomic_ops.
  -:  --------- >  48:  04914cfbc Misc: moved documentation in generated ZIP archive.
  -:  --------- >  49:  22a2a225b Added "keepalive_min_timeout" directive.
  -:  --------- >  50:  46b9f5d38 SNI: added restriction for TLSv1.3 cross-SNI session resumption.
  -:  --------- >  51:  ecb809305 nginx-1.27.4-RELEASE
  -:  --------- >  52:  f274b3f72 Version bump.
  -:  --------- >  53:  9a4090f02 Core: fix build without libcrypt.
  -:  --------- >  54:  3327353ec Configure: MSVC compatibility with PCRE2 10.45.
  -:  --------- >  55:  f51e2de6f Add gitignore file.
  -:  --------- >  56:  d25139db0 Improved ngx_http_subrequest() error handling.
  -:  --------- >  57:  b11ae4cfc SSL: style.
  -:  --------- >  58:  3d7304b52 SSL: using static storage for NGX_SSL_MAX_SESSION_SIZE buffers.
  -:  --------- >  59:  912459220 SSL: raised limit for sessions stored in shared memory.
  -:  --------- >  60:  311c39037 SSL: improved logging of saving sessions from upstream servers.
  -:  --------- >  61:  d16251969 SSL: removed stale comments.
  -:  --------- >  62:  d31305653 Slice filter: improved memory allocation error handling.
  -:  --------- >  63:  a813c6392 Charset filter: improved validation of charset_map with utf-8.
  -:  --------- >  64:  6c3a9d561 Upstream: fixed passwords support for dynamic certificates.
  -:  --------- >  65:  b6e7eb0f5 SSL: external groups support in $ssl_curve and $ssl_curves.
  -:  --------- >  66:  2b8b70068 QUIC: graph-friendly congestion control logging.
  -:  --------- >  67:  3a97111ad HTTP/3: graceful shutdown on keepalive timeout expiration.
  -:  --------- >  68:  53e7e9eb5 QUIC: use path MTU in congestion window computations.
  -:  --------- >  69:  38236bf74 QUIC: prevent spurious congestion control recovery mode.
  -:  --------- >  70:  1e883a40d QUIC: ngx_msec_t overflow protection.
  -:  --------- >  71:  04c65ccd9 QUIC: all-levels commit and revert functions.
  -:  --------- >  72:  cd5e4fa14 QUIC: do not increase underutilized congestion window.
  -:  --------- >  73:  6bf13e9d5 QUIC: do not shrink congestion window after losing an MTU probe.
  -:  --------- >  74:  a40cc7002 QUIC: ignore congestion control when sending MTU probes.
  -:  --------- >  75:  f9a7e7cc1 QUIC: CUBIC congestion control.
  -:  --------- >  76:  2fb32ff24 QUIC: optimized connection frame threshold.
  -:  --------- >  77:  aa49a416b QUIC: dynamic packet threshold.
  -:  --------- >  78:  6ac8b69f0 nginx-1.27.5-RELEASE
  -:  --------- >  79:  0626e60a7 Version bump.
  -:  --------- >  80:  04813dac8 QUIC: lowered log level for unsupported transport parameters.
  -:  --------- >  81:  444954aba Fixed -Wunterminated-string-initialization with gcc15.
  -:  --------- >  82:  0f9f43b79 HTTP/3: fixed NGX_HTTP_V3_VARLEN_INT_LEN value.
  -:  --------- >  83:  b9d0ba667 Core: improved NGX_ALIGNMENT detection on some x86_64 platforms.
  -:  --------- >  84:  020b1db7e Win32: added detection of ARM64 target.
  -:  --------- >  85:  9785db9bd QUIC: fixed a typo.
  -:  --------- >  86:  4f8bc0b28 SSL: fixed build with OPENSSL_NO_DEPRECATED.
  -:  --------- >  87:  adda70415 SSL: fixed build with OPENSSL_NO_DH.
  -:  --------- >  88:  f3542500b QUIC: do not block ACKs by congestion control.
  -:  --------- >  89:  335993154 QUIC: removed level field from ngx_quic_compat_record_t.
  -:  --------- >  90:  39b1e3fe9 QUIC: removed excessive casts for ngx_ssl_get_connection().
  -:  --------- >  91:  aa43385ff QUIC: removed ALPN feature test.
  -:  --------- >  92:  ef9cd3214 QUIC: logging level of handshake errors.
  -:  --------- >  93:  47f96993f QUIC: logging of SSL library errors.
  -:  --------- >  94:  7468a10b6 QUIC: adjusted handling of callback errors.
  -:  --------- >  95:  5d7fd4a7e QUIC: reset qc->error to zero again.
  -:  --------- >  96:  54e6b7cfe QUIC: logging missing mandatory TLS extensions only once.
  -:  --------- >  97:  e561f7dbc QUIC: defined SSL API macros in a single place.
  -:  --------- >  98:  9857578f1 QUIC: factored out SSL_provide_quic_data() to the helper function.
  -:  --------- >  99:  bcb9d3fd2 QUIC: ssl_encryption_level_t abstraction layer.
  -:  --------- > 100:  1d4d2f2c9 QUIC: better approach for premature handshake completion.
  -:  --------- > 101:  6a134dfd4 QUIC: using QUIC API introduced in OpenSSL 3.5.
  -:  --------- > 102:  0fdbfc1ff SSL: support loading keys via OSSL_STORE.
  -:  --------- > 103:  3d5889a3e SSL: disabled UI console prompts from worker processes.
  -:  --------- > 104:  5b8a5c08c Core: added support for TCP keepalive parameters on macOS.
  -:  --------- > 105:  ea001feb1 HTTP/2: added function declaration.
  -:  --------- > 106:  662c1dd2a Upstream: early hints support.
  -:  --------- > 107:  ba917b136 HTTP/3: indexed field line encoding for "103 Early Hints".
  -:  --------- > 108:  c370ac8a5 Use NGX_CONF_OK in some function return checks.
  -:  --------- > 109:  4eaecc5e8 Use NULL instead of 0 for null pointer constant.
  -:  --------- > 110:  cdf7a9c6c Upstream: fixed reinit request with gRPC and Early Hints.
  -:  --------- > 111:  cedb855d7 QUIC: disabled OpenSSL 3.5 QUIC API support by default.
  -:  --------- > 112:  b997be14f Win32: skip OpenSSL dependency generation to conserve time.
  -:  --------- > 113:  1263d6bec Updated OpenSSL and PCRE used for win32 builds.
  -:  --------- > 114:  235f40990 nginx-1.29.0-RELEASE
  -:  --------- > 115:  279fe352c Version bump.
  -:  --------- > 116:  d1843e1d9 Win32: fixed PCRE license for nginx/Windows zip.
  -:  --------- > 117:  0bb7489cb QUIC: adjusted OpenSSL 3.5 QUIC API feature test.
  -:  --------- > 118:  a5d60c30d SSL: SSL_group_to_name() compatibility macro.
  -:  --------- > 119:  a5ca38f30 SSL: fixed testing OPENSSL_VERSION_NUMBER for OpenSSL 3.0+.
  -:  --------- > 120:  0daaba5c5 Events: fixed -Wzero-as-null-pointer-constant warnings in kqueue.
  -:  --------- > 121:  3f5f8a7f5 Configure: set NGX_KQUEUE_UDATA_T at compile time.
  -:  --------- > 122:  c52c5698c Events: compatibility with NetBSD 10.0 in kqueue.
  -:  --------- > 123:  3739fe94d HTTP/3: fixed potential type overflow in string literal parser.
  -:  --------- > 124:  4da771108 HTTP/3: fixed handling of :authority and Host with port.
  -:  --------- > 125:  50932c3c6 HTTP/2: fixed flushing early hints over SSL.
  -:  --------- > 126:  a238bb3d2 HTTP/2: factored out constructing the Host header.
  -:  --------- > 127:  ede5623b1 HTTP/2: fixed handling of the ":authority" header.
  -:  --------- > 128:  4d857aaf4 Made ngx_http_process_request_header() static again.
  -:  --------- > 129:  372659114 HTTP/3: improved invalid ":authority" error message.
  -:  --------- > 130:  f4005126d Updated ngx_http_process_multi_header_lines() comments.
  -:  --------- > 131:  ed99269ee SSL: disabled certificate compression by default with OpenSSL.
  -:  --------- > 132:  251444fcf SSL: support for compressed server certificates with OpenSSL.
  -:  --------- > 133:  034f15bbc Auth basic: fixed file descriptor leak on memory allocation error.
  -:  --------- > 134:  765642b86 Mail: improved error handling in plain/login/cram-md5 auth methods.
  -:  --------- > 135:  9c02c84a7 Mail: reset stale auth credentials with "smtp_auth none;".
  -:  --------- > 136:  239e10793 Mail: logging upstream to the error log with "smtp_auth none;".
  -:  --------- > 137:  cc1c07ca3 Updated OpenSSL used for win32 builds.
  -:  --------- > 138:  0024724f2 nginx-1.29.1-RELEASE
  -:  --------- > 139:  36d40e561 Version bump.
  -:  --------- > 140:  1a82df8cc Added a previously missed changes entry in 1.29.1 relnotes.
  -:  --------- > 141:  446ce033e Removed legacy charset directive from default config example.
  -:  --------- > 142:  eb5ebbbed QUIC: fixed ssl_reject_handshake error handling.
  -:  --------- > 143:  417c87b78 Updated link to xslscript.
  -:  --------- > 144:  bc71625dc Fixed inaccurate index directive error report.
  -:  --------- > 145:  0373fe5d9 SNI: using the ClientHello callback.
  -:  --------- > 146:  7f9ced0ce SNI: support for early ClientHello callback with BoringSSL.
  -:  --------- > 147:  4c9ae11df QUIC: localized OpenSSL headers used for QUIC protection.
  -:  --------- > 148:  af436c58c QUIC: a new macro to differentiate BoringSSL specific EVP API.
  -:  --------- > 149:  93ff1ee12 SSL: AWS-LC compatibility.
  -:  --------- > 150:  8255bd29a Upstream: overflow detection in Cache-Control delta-seconds.
  -:  --------- > 151:  6f81314a0 Mail: xtext encoding (RFC 3461) in XCLIENT LOGIN.
  -:  --------- > 152:  c2a266fa7 Added F5 CLA workflow.
  -:  --------- > 153:  a144d828c SSL: fixed "key values mismatch" with object cache inheritance.
  -:  --------- > 154:  9d5cdc616 Fixed a typo.
  -:  --------- > 155:  7f71abdd1 nginx-1.29.2-RELEASE
  -:  --------- > 156:  3494f25c3 Version bump.
  -:  --------- > 157:  f5a989cda Updated OpenSSL and PCRE used for win32 builds.
  -:  --------- > 158:  25b03d650 SSL: disabled using certificate compression with OCSP stapling.
  -:  --------- > 159:  78d1ab5a2 SSL: support for compressed server certificates with BoringSSL.
  -:  --------- > 160:  c8c7beb96 Added $request_port and $is_request_port variables.
  -:  --------- > 161:  42ca3a457 CONNECT method support for HTTP/1.1.
  -:  --------- > 162:  364a94ece Upstream: reset local address in case of error.
  -:  --------- > 163:  71f8eb52b SSL: $ssl_sigalg, $ssl_client_sigalg.
  -:  --------- > 164:  ac72ca60c Geo: the "volatile" parameter.
  -:  --------- > 165:  c93a0c48a Headers filter: inheritance control for add_header and add_trailer.
  -:  --------- > 166:  ce30a1cb0 OCSP: fixed invalid type for the 'ssl_ocsp' directive.
  -:  --------- > 167:  f04e2b7f6 Fixed compilation warnings on Windows after c93a0c48af87.
  -:  --------- > 168:  65c0b2e77 Modules compatibility: increased compat section size.
  -:  --------- > 169:  fd0848bdd nginx-1.29.3-RELEASE
  -:  --------- > 170:  99312be10 Configure: ensure we get the "built by ..." line in nginx -V.
  -:  --------- > 171:  a386df6b2 Version bump.
  -:  --------- > 172:  ac99f2808 Configure: MSVC compatibility with PCRE2 10.47.
  -:  --------- > 173:  38a701d88 SSL: ngx_ssl_set_client_hello_callback() error handling.
  -:  --------- > 174:  9d04b6630 SSL: fixed build with BoringSSL, broken by 38a701d88.
  -:  --------- > 175:  6ed118841 HTTP/2: extended guard for NULL buffer and zero length.
  -:  --------- > 176:  511abb19e Improved host header validation.
  -:  --------- > 177:  6446f9910 Changed interface of ngx_http_validate_host().
  -:  --------- > 178:  bcb41c919 Proxy: fixed segfault in URI change.
  -:  --------- > 179:  ab4f5b2d3 Add basic ECH shared-mode via OpenSSL.
  1:  a42220319 = 180:  e19e7a809 Add HTTP_HOST to fastcgi params
```

ac000 commented:
* This version uses the new `$is_request_port` & `$request_port`
  variables and adds this new parameter to more files.
* The 'Author' has been switched due to the CLA, but the original author
  has been added as a `Co-authored-by:` commit tag so they still get
  credit.

```
$ git range-diff --creation-factor=100  e19e7a809...d39c9b0cf
1:  e19e7a809 ! 1:  d39c9b0cf Add HTTP_HOST to fastcgi params
    @@
      ## Metadata ##
    -Author: Thomas Ward <teward@thomas-ward.net>
    +Author: Andrew Clayton <a.clayton@nginx.com>
     
      ## Commit message ##
    -    Add HTTP_HOST to fastcgi params
    +    Add HTTP_HOST parameter.
     
    -    https://salsa.debian.org/nginx-team/nginx/-/issues/21
    +    This ensures that a HTTP_HOST environment variable is always set
    +    regardless of HTTP protocol version used.
    +
    +    Link: <https://salsa.debian.org/nginx-team/nginx/-/issues/21>
    +    Co-authored-by: Thomas Ward <teward@thomas-ward.net>
    +
    + ## conf/fastcgi.conf ##
    +@@ conf/fastcgi.conf: fastcgi_param  SERVER_ADDR        $server_addr;
    + fastcgi_param  SERVER_PORT        $server_port;
    + fastcgi_param  SERVER_NAME        $server_name;
    + 
    ++fastcgi_param  HTTP_HOST          $host$is_request_port$request_port;
    ++
    + # PHP only, required if PHP was built with --enable-force-cgi-redirect
    + fastcgi_param  REDIRECT_STATUS    200;
     
      ## conf/fastcgi_params ##
     @@ conf/fastcgi_params: fastcgi_param  SERVER_ADDR        $server_addr;
      fastcgi_param  SERVER_PORT        $server_port;
      fastcgi_param  SERVER_NAME        $server_name;
      
    -+fastcgi_param  HTTP_HOST          $host;
    ++fastcgi_param  HTTP_HOST          $host$is_request_port$request_port;
     +
      # PHP only, required if PHP was built with --enable-force-cgi-redirect
      fastcgi_param  REDIRECT_STATUS    200;
    +
    + ## conf/scgi_params ##
    +@@ conf/scgi_params: scgi_param  REMOTE_ADDR        $remote_addr;
    + scgi_param  REMOTE_PORT        $remote_port;
    + scgi_param  SERVER_PORT        $server_port;
    + scgi_param  SERVER_NAME        $server_name;
    ++
    ++scgi_param  HTTP_HOST          $host$is_request_port$request_port;
    +
    + ## conf/uwsgi_params ##
    +@@ conf/uwsgi_params: uwsgi_param  REMOTE_ADDR        $remote_addr;
    + uwsgi_param  REMOTE_PORT        $remote_port;
    + uwsgi_param  SERVER_PORT        $server_port;
    + uwsgi_param  SERVER_NAME        $server_name;
    ++
    ++uwsgi_param  HTTP_HOST          $host$is_request_port$request_port;
```

pluknet commented:
I'll take a look, tnx.

pluknet commented:
@ac000 

I've been pondering on how to best fix this without introducing
potentially breaking changes.

The issue is that HTTP_HOST is quite commonly redefined in nginx
configuration via [fastcgi_param](http://nginx.org/r/fastcgi_param), with conf/fastcgi_params included,
following the commented out FastCGI example in conf/nginx.conf.

Taking this into account, updating conf/fastcgi_params (and friends)
to include HTTP_HOST would ultimately result in multiple
"fastcgi_param HTTP_HOST" directives.  This introduces the problem
of passing HTTP_HOST twice.
This can be unnoticed on upgrade because fastcgi_param doesn't try
to detect such collisions (generally it's the user's responsibility).
Passing multiple parameters with the same name is explicitly forbidden
in CGI specification (for obvious reasons).  In the worst case, this can
result in a silent breakage, due to host confusion, or in denial of service,
due to invalid FastCGI requests, depending on implementation details of
the relevant protocol on backend.

So what I think we can do here is to come back to the previous approach
with changing [fastcgi_param](http://nginx.org/r/fastcgi_param) defaults to include HTTP_HOST there, to pass
it as ngx_http_fastcgi_headers[] in ngx_http_fastcgi_init_params(), similarly in other modules.

The fastcgi_param directives (params_source) have a priority over
default_params when merging parameters in ngx_http_fastcgi_init_params(),
so it is not a problem if user has its own HTTP_HOST configuration.
Also, the CGI spec statement regarding "protocol-specific meta-variables"
doesn't seem to contradict this new behaviour, specifically:
: The server SHOULD set meta-variables specific to the protocol
IMHO, setting HTTP_HOST to ":authority" is such an alignment to this statement.

For the record, previously suggested to add HTTP_HOST to conf/fastcgi_params
instead would allow to avoid changing the well established logic of passing client
request headers as "HTTP_" params in modules, and would allow to avoid potentially
breaking existing configurations (but see above for alignment), with more flexibility
to alter it as needed. This seems to not work.

ac000 commented:
@pluknet

OK. I'll close this pull-request and put up a new one.
